libevmasm/Assembly.cpp

```diff
@@ -297,6 +503,142 @@ Json::Value Assembly::assemblyJSON(std::map<std::string, unsigned> const& _sourc
 	return root;
 }
 
+std::pair<std::shared_ptr<Assembly>, std::vector<std::string>> Assembly::fromJSON(
+	Json::Value const& _json,
+	std::vector<std::string> const& _sourceList,
+	size_t _level
+)
+{
+	solRequire(_json.isObject(), AssemblyImportException, "Supplied JSON is not an object.");
+	static std::set<std::string> const validMembers{".code", ".data", ".auxdata", "sourceList"};
+	for (std::string const& attribute: _json.getMemberNames())
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+
+	if (_level == 0)
+	{
+		if (_json.isMember("sourceList"))
+		{
+			solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
+			for (Json::Value const& sourceName: _json["sourceList"])
+				solRequire(sourceName.isString(), AssemblyImportException, "The 'sourceList' array contains an item that is not a string.");
+		}
+	}
+	else
+		solRequire(
+			!_json.isMember("sourceList"),
+			AssemblyImportException,
+			"Member 'sourceList' may only be present in the root JSON object."
+		);
+
+	auto result = std::make_shared<Assembly>(EVMVersion{}, _level == 0 /* _creation */, "" /* _name */);
+	std::vector<std::string> parsedSourceList;
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		solAssert(_sourceList.empty());
+		for (Json::Value const& sourceName: _json["sourceList"])
+		{
+			solRequire(
+				std::find(parsedSourceList.begin(), parsedSourceList.end(), sourceName.asString()) == parsedSourceList.end(),
```

Comment:
This is basically just a nested for loop - so a bit wasteful; wouldn't it make more sense to make `parsedSourceList` a `std::set<std::string>` and use `parsedSourceSet.count(sourceName.asString())` instead, and then just copy it to a vector when/if needed, i.e. `std::vector<std::string> parsedSourceList(parsedSourceSet.begin(), parsedSourceSet.end())`?

Reply:
Wouldn't that destroy the order though? The order here is very important.

But other than that, yeah, that would be more efficient for long lists.

Reply:
Yes, it would, in which case we couldn't simply copy content over from the set into the vector, but could none the less use the same approach (i.e. using a set for a presence check, which technically wouldn't be any less efficient.

Reply:
In that case it would be better to first load all items into a vector and then only use the set as a temporary helper to check if items are unique. Even better, wrap that in a reusable `isUnique()` helper. Or maybe such a helper even already exists in boost?

But IMO this is also good enough as is, given that the feature is experimental and we gave some other inefficiencies and suboptimal things a pass too. This could be improved to the follow-up refactor PR.